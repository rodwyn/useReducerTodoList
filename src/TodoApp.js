import { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import "./App.css";
import { useForm } from "./hooks/useForm";
import { TodoList } from "./components/TodoList";

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);
  
  const [{description}, handleInputChange, reset] = useForm({
    description: ''
  });

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = todoId => {
    dispatch({
      type: 'delete',
      payload: todoId
    });
  };

  const handleToggle = todoId => {
    dispatch({
      type: 'toggle',
      payload: todoId
    });
  };
  
  const handleSubmit = event => {
    event.preventDefault();
    
    if (description.trim().length < 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    };

    const action = {
      type: 'add',
      payload: newTodo
    };

    dispatch(action);
    reset();
  };
  
  return (
    <div className="App">
      <h1>Todo App <small>({todos.length})</small></h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <h4>Add ToDo</h4>
          <hr />
          <form onSubmit={ handleSubmit }>
            <input
              type="text"
              name="description"
              placeholder="Add Todo"
              autoComplete="off"
              className="form-control"
              onChange={handleInputChange}
              value={description}
               />
            <div className="d-grid gap-2 mt-2">
              <button className="btn btn-outline-primary" type="submit">Add</button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
