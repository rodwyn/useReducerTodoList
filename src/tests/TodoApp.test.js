import React from 'react';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';
import { TodoApp } from '../TodoApp';
import { demosTodos } from './fixtures/demosTodos';

describe('Testing <TodoApp /> component.', () => {
  const wrapper = shallow(<TodoApp />);

  test('should render correctly.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should add a Todo.', () => {
    const wrapper = mount(<TodoApp />);

    act(()=> {
      wrapper.find('TodoForm').prop('handleAddTodo')(demosTodos[0]);
    });


    expect(wrapper.find('h1').text().trim()).toBe('Todo App (1)');
  });
});
