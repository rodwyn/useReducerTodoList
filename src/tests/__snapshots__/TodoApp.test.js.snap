// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TodoApp /> component. should render correctly. 1`] = `
<div
  className="App"
>
  <h1>
    Todo App 
    <small>
      (
      0
      )
    </small>
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
